cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/Dag.java

```diff
@@ -0,0 +1,403 @@
+    }
+
+    // check for cycles
+    linearize();
+
+    // check for sections of the dag that are on an island by themselves
+
+    // source -> [ nodes accessible by the source ]
+    Map<String, Set<String>> nodesAccessibleBySources = new HashMap<>();
+    for (String source : sources) {
+      nodesAccessibleBySources.put(source, accessibleFrom(source));
+    }
+
+    // the algorithm is to keep an island and try to keep adding to it until we can't anymore.
+    // the island starts off as the nodes accessible by the first source
+    // we then loop through all other sources and add them to the island if they can access any node in the island.
+    // we stop if we ever loop through all other sources and can't add them to the island,
+    // or if the island has grown to include all sources.
+    Set<String> islandNodes = new HashSet<>();
+    // seed the island with the first source
+    Set<String> potentialIslandSources = new HashSet<>(sources);
+    String firstSource = potentialIslandSources.iterator().next();
+    islandNodes.addAll(nodesAccessibleBySources.get(firstSource));
+    potentialIslandSources.remove(firstSource);
+
+    while (!potentialIslandSources.isEmpty()) {
+      Set<String> sourcesAdded = new HashSet<>();
+      // for each source not yet a part of the island
+      for (String potentialIslandSource : potentialIslandSources) {
+        Set<String> accessibleBySource = nodesAccessibleBySources.get(potentialIslandSource);
+        // if that source can access the island in any way, add it to the island
+        if (!Sets.intersection(islandNodes, accessibleBySource).isEmpty()) {
+          islandNodes.addAll(accessibleBySource);
+          sourcesAdded.add(potentialIslandSource);
+        }
+      }
+      // if this is empty, no sources were added to the island. That means the island really is an island.
+      if (sourcesAdded.isEmpty()) {
+        throw new IllegalStateException(
+          String.format("Invalid DAG. Stages %s form an island (no other stages connect to them).",
+                        Joiner.on(',').join(islandNodes)));
+      }
+      potentialIslandSources.removeAll(sourcesAdded);
+    }
+  }
+
+  public Set<String> getNodes() {
+    return nodes;
+  }
+
+  public Set<String> getSources() {
+    return Collections.unmodifiableSet(sources);
+  }
+
+  public Set<String> getSinks() {
+    return Collections.unmodifiableSet(sinks);
+  }
+
+  public Set<String> getNodeOutputs(String node) {
+    return Collections.unmodifiableSet(outgoingConnections.get(node));
+  }
+
+  public Set<String> getNodeInputs(String node) {
+    return Collections.unmodifiableSet(incomingConnections.get(node));
+  }
+
+  /**
+   * Return all stages accessible from the specified stage.
+   *
+   * @param stage the stage to start at
+   * @return all stages accessible from that stage
+   */
+  public Set<String> accessibleFrom(String stage) {
```

Comment:
this will be used later to help calculate where we need to insert a local dataset to connect two mapreduce jobs